### PR TITLE
Feature/system tracer

### LINKED ISSUE #90: Smaller Serial logging API (prints)
**Summary:** 

It will be extremely beneficial if we add a smaller printf (UART print serial) library to the BSP  

**Why do we need this?**

To be able to use an OS and a larger applications while we preserve the capability to do serial logging 

**What is already there? What do you see now?**

We have a very good (but big) advanced tracer API built over the ST utilities, that is useful in small bare-metal projects

**What is missing? What do you want to see?**

A smaller print API that allows for simple logging in bigger more memory demanding applications.

**How do you propose to implement this?**

- Look into ST UART examples and find the smallest implementation
- Integrate that implementation into the BSP and the HAL layer
- Add a configuration parameter to allow for choice between smaller and bigger logging implementations

**Environment:**

OS and Bare-metal

**Acceptance Criteria:**

- Should compile all apps with no issue
- Should be smaller in size than the current implementation of the advanced tracer

**What can you do yourself and what do you need help with?**

ALL

### PR DESCRIPTION
**Pull request by elsalahy**
**Summary:**

This update adds a system wide advanced and tiny tracer that can be used according to application need.
Apps like basic_bootloader and freertos(basic, LoRaWAN) can use tiny tracer to reduce overhead.
Apps like basic and basic_lorawan can use Advancer tracer 

Closes #90 

Also added a minor fix to freertos_lorawan app when testing the new tracer, relates to #88 and #99

**Changes:**

- Add GNSE Tracer
- Adapt libs and apps to use the new tracer

**Notes for Reviewers:**

I ran my tests, but I could have overlooked something.
Please give it a run and test random apps and report if you faced any issue compiling or observing a trace.

Will update the cmake_dev (faster cmakes) when the PR is approved, to avoid multiple change with reviews.

**Release Notes: _(optional)_**

...

### CODE REVIEW

File: Software/app/freefall_lorawan/sys_app.c [hunk 1]
```diff
@@ -73,12 +73,8 @@ void SystemApp_Init(void)
   /* Configure the debug mode*/
   DBG_Init();
 
-  /*Initialize the terminal */
```
Comment (marnixcro) [inline Software/app/freefall_lorawan/sys_app.c:76]: Maybe keep this line (with added space)?

File: Software/app/freertos_lorawan/sys_app.c [hunk 2]
```diff
@@ -67,12 +55,7 @@ void SystemApp_Init(void)
   /* Configure the debug mode*/
   DBG_Init();
 
-  /*Initialize the terminal */
```
Comment (marnixcro) [inline Software/app/freertos_lorawan/sys_app.c:70]: Maybe keep this line (with added space)?

File: Software/app/secure_element_lorawan/sys_app.c [hunk 3]
```diff
@@ -72,11 +72,8 @@ void SystemApp_Init(void)
   DBG_Init();
 
   /*Initialize the terminal */
```
Comment (marnixcro) [inline Software/app/secure_element_lorawan/sys_app.c:74]: Line is kept here (on others its removed), but maybe add a small space here?

**Review by marnixcro — APPROVED** [review]
Nice changes. I could build all apps on cubeide and cmake, no errors. All functions seem to work. I also checked for any dangling app_log_enable references. I only have a few minor optional changes you might want to fix.

File: Software/app/basic_lorawan/conf/app_conf.h [hunk 4]
```diff
@@ -24,7 +24,7 @@
 #ifndef APP_CONF_H
 #define APP_CONF_H
 
-#define APP_LOG_ENABLED 1
+#define GNSE_ADVANCED_TRACER_ENABLE 1
 
 /* if ON (=1) it enables the debugger plus 4 dgb pins */
```
Comment (marnixcro) [inline Software/app/basic_lorawan/conf/app_conf.h:29]: More unrelated spelling: dbg

File: Software/app/freertos_lorawan/conf/app_conf.h [hunk 5]
```diff
@@ -26,7 +26,7 @@
 
 #include "lorawan_conf.h"
 
-#define APP_LOG_ENABLED 1
+#define GNSE_TINY_TRACER_ENABLE 1
 
 /* if ON (=1) it enables the debugger plus 4 dgb pins */
```
Comment (marnixcro) [inline Software/app/freertos_lorawan/conf/app_conf.h:31]: More unrelated spelling: dbg

File: Software/app/app.h [hunk 6]
```diff
@@ -25,11 +25,10 @@
 #define APP_H
 
 /**
- * Utilities
+ * Standered libraries
```
[the remark this answers is not in the document]
Comment (elsalahy) [reply Software/app/app.h:28]: Done

File: Software/app/basic_lorawan/lora_app.c [hunk 7]
```diff
@@ -223,10 +223,10 @@ static void OnRxData(LmHandlerAppData_t *appData, LmHandlerRxParams_t *params)
       }
       break;
     case LORAWAN_APP_PORT:
-      APP_LOG(TS_OFF, VLEVEL_M, "\r\n Recieved %d bytes on LORAWAN_APP_PORT: %d \r\n", appData->BufferSize, LORAWAN_APP_PORT);
+      APP_LOG(ADV_TRACER_TS_OFF, ADV_TRACER_VLEVEL_M, "\r\n Recieved %d bytes on LORAWAN_APP_PORT: %d \r\n", appData->BufferSize, LORAWAN_APP_PORT);
       break;
     default:
-      APP_LOG(TS_OFF, VLEVEL_M, "\r\n Recieved %d bytes on undefined port: %d \r\n", appData->BufferSize, LORAWAN_APP_PORT);
+      APP_LOG(ADV_TRACER_TS_OFF, ADV_TRACER_VLEVEL_M, "\r\n Recieved %d bytes on undefined port: %d \r\n", appData->BufferSize, LORAWAN_APP_PORT);
```
[the remark this answers is not in the document]
Comment (elsalahy) [reply Software/app/basic_lorawan/lora_app.c:229]: Done

File: Software/app/secure_element_lorawan/lora_app.c [hunk 8]
```diff
@@ -136,7 +136,7 @@ static void OnRxData(LmHandlerAppData_t *appData, LmHandlerRxParams_t *params)
 {
   if ((appData != NULL) && (params != NULL))
   {
-    APP_LOG(TS_OFF, VLEVEL_M, "\r\n Recieved Unhandled Downlink on F_PORT:%d \r\n", appData->Port);
+    APP_LOG(ADV_TRACER_TS_OFF, ADV_TRACER_VLEVEL_M, "\r\n Recieved Unhandled Downlink on F_PORT:%d \r\n", appData->Port);
```
[the remark this answers is not in the document]
Comment (elsalahy) [reply Software/app/secure_element_lorawan/lora_app.c:139]: Done

File: Software/lib/MX25R1635/mx_define.h [hunk 9]
```diff
@@ -26,12 +26,11 @@
...
+#include "GNSE_tracer.h"
 #include "GNSE_bsp.h"
 
 #define GetChar()
-#define Mx_printf FLASH_PPRINTF
-#define FLASH_PPRINTF(...)  do{ } while( UTIL_ADV_TRACE_OK != UTIL_ADV_TRACE_COND_FSend(VLEVEL_OFF, T_REG_OFF, TS_OFF, __VA_ARGS__) ) //Polling Mode
+#define Mx_printf LIB_PRINTF
 #define MxTime uint32_t
 #define MxGetTime(TimeVal) do{ {TimeVal = HAL_GetTick();}} while(0);
 #define COUNTS_PER_SECOND 1000 //Beacuse HAL_GetTick() returns ms vlaues and the driver uses uS
```
[the remark this answers is not in the document]
Comment (elsalahy) [reply Software/lib/MX25R1635/mx_define.h:36]: Done

File: Software/app/basic/conf/app_conf.h [hunk 10]
```diff
@@ -24,7 +24,7 @@
 #ifndef APP_CONF_H
 #define APP_CONF_H
 
-#define APP_LOG_ENABLED 1
+#define GNSE_ADVANCED_TRACER_ENABLE 1
 
 /* if ON (=1) it enables the debbugger plus 4 dgb pins */
```
[the remark this answers is not in the document]
Comment (elsalahy) [reply Software/app/basic/conf/app_conf.h:29]: Done

File: Software/app/freefall_lorawan/conf/app_conf.h [hunk 11]
```diff
@@ -24,7 +24,7 @@
 #ifndef APP_CONF_H
 #define APP_CONF_H
 
-#define APP_LOG_ENABLED 1
+#define GNSE_ADVANCED_TRACER_ENABLE 1
 
 /* if ON (=1) it enables the debbugger plus 4 dgb pins */
```
[the remark this answers is not in the document]
Comment (elsalahy) [reply Software/app/freefall_lorawan/conf/app_conf.h:29]: Done

[the remark this answers is not in the document]
Comment (elsalahy) [reply Software/app/freefall_lorawan/conf/app_conf.h:29]: Done

File: Software/app/secure_element_lorawan/conf/app_conf.h [hunk 12]
```diff
@@ -24,7 +24,7 @@
 #ifndef APP_CONF_H
 #define APP_CONF_H
 
-#define APP_LOG_ENABLED 1
+#define GNSE_ADVANCED_TRACER_ENABLE 1
 
 /* if ON (=1) it enables the debugger plus 4 dgb pins */
```
[the remark this answers is not in the document]
Comment (elsalahy) [reply Software/app/secure_element_lorawan/conf/app_conf.h:29]: Done

File: Software/app/freefall_lorawan/sys_app.c (hunk 1 above)
Comment (elsalahy) [reply Software/app/freefall_lorawan/sys_app.c:76]: Done

File: Software/app/freertos_lorawan/sys_app.c (hunk 2 above)
Comment (elsalahy) [reply Software/app/freertos_lorawan/sys_app.c:70]: Done

File: Software/app/secure_element_lorawan/sys_app.c (hunk 3 above)
Comment (elsalahy) [reply Software/app/secure_element_lorawan/sys_app.c:74]: Done

File: Software/app/basic_lorawan/conf/app_conf.h (hunk 4 above)
Comment (elsalahy) [reply Software/app/basic_lorawan/conf/app_conf.h:29]: Done

File: Software/app/freertos_lorawan/conf/app_conf.h (hunk 5 above)
Comment (elsalahy) [reply Software/app/freertos_lorawan/conf/app_conf.h:31]: Done

**Comment by elsalahy** [comment]
@marnixcro Thanks for the review, I will not update tools/cmake_dev now and I think we need to come up with a script that automates the generation of these files

**Review by marnixcro — APPROVED** [review]

**Pull request merged**